Refactor model updates

jcaamano opened the pull request:
This commit is a huge refactor of how model updates are handled. Through
a new updates package and ModelUpdates type, abstracted to a generic
database Update type, updates coming both from OVSDB operations (server)
and monitor notifications (client) are handled with a common code base,
along with the following benefits:

* A single place to handle updates also lays down a clear path of where
  and how validation and model or row cloning should happen which can be
  useful for performance optimization changes in the future.

* Considerably reduced the number of panics in the server where we
  now return proper error operation results.

* The server is also now able to aggregate correctly several operations
  for the same row in a single transaction.

* The server also does not build an updated model twice per operation,
  the same updated model is now used when applied to the transaction cache
  as well as the database cache.

This PR is a pre-requisite to be able to implement referential integrity reasonably well.

Signed-off-by: Jaime Caamaño Ruiz <jcaamano@redhat.com>

hzhou8 commented:
Hi @jcaamano , thanks for the improvements. Could you explain in more detail about the packages created and the motivation?

I just looked at the https://github.com/ovn-org/libovsdb/pull/340, which moved several files of the server package to a new database package, but I didn't understand why was it necessary. It seems those files are still related to server side implementation, so what's the benefit of moving them to a new database package?

For the current PR, I got more confused. I haven't looked into deeper yet, but the description claims that it is to reuse common code, yet I am seeing 2000+ code than before with this PR. I appreciate if you could explain your long term plan more clearly to understand such big changes.

At the same time, it would be good to update the documentation for the new packages: https://github.com/ovn-org/libovsdb#documentation

Thank you!

jcaamano commented:
> Hi @jcaamano , thanks for the improvements. Could you explain in more detail about the packages created and the motivation?
> 
> I just looked at the #340, which moved several files of the server package to a new database package, but I didn't understand why was it necessary. It seems those files are still related to server side implementation, so what's the benefit of moving them to a new database package?

- For the database package introduced in #340 , there is a Database interface which the transaction and the server are prepared to use. It does not make sense to have a database implementation in the same package as the server (nor it makes sense to have the database implementation in the same package as the database interface, which I changed in this PR). Someone might want to have a server with a different database implementation.

- For the transaction package, we have been considering using it in the client as a validating throw-away transaction, where operations are validated against the client cache before being sent towards the server. Also considering the fact that if someone would have it's own database implementation, it may or may not have it's own transaction implementation as well depending on the desired or needed level of coupling of the transaction with the database. 
But we can move this back to the server package for now I guess (let me know if this would be preferred for now) but it was just as easy for me to move it to it's own package.

- The updates package makes sense as model updates are not only handled only in the context of a transaction, but also as Update2 modifications are received client side. Previously we had duplicated code in server and client regarding these updates (ApplyModifications).

- I don't think that the fact that stuff is only used in the server warrants for everything to be in the same server package anyway. Having separate packages and clear interfaces facilitates reuse and testing so it is nice to push towards an out-of-alpha implementation state for the server as we make changes. Sure, we could actually wait until there is an actual need but it just seemed like an improvement to me in any case so I went for it.

All in all, I think they are reasonable decisions. Happy to get specific suggestions on how things could be organized better though.

> For the current PR, I got more confused. I haven't looked into deeper yet, but the description claims that it is to reuse common code, yet I am seeing 2000+ code than before with this PR. I appreciate if you could explain your long term plan more clearly to understand such big changes.

Right, I claimed code reuse, not less code. This added code is contributing to facilitate reuse, currently and for future changes; to fix stuff that was not working correctly on the server side currently; to avoid doing stuff twice at runtime (an updated model is only calculated once, both for the transaction cache and the database cache, instead of twice) probably improving server performance on the way. A single approach converged to being able to to rip all those benefits. But of the 2000+ extra lines, much of them can be attributed to more unit tests as is usually the case. For example, consider that the new updates package has 2500+ lines worth of new tests.

>  I appreciate if you could explain your long term plan more clearly to understand such big changes.

I guess this could have been divided in smaller PRs but the important functional gaps that the server has and the review pace on this project actually favors covering them in less bigger PRs rather than more smaller PRs.

The short-term plan is to have the server code in a state where we can implement constraint validations reasonably well: we fixed index constraint validation. We want to implement referential integrity validation, maxRows validation, min/max carnality and other column data type validations, etc. We need these working for testing in downstream projects.
The long-term plan is to have the server out of alpha state functionally. 
The very-long-term plan might be to have the server out of alpha state performance-wise. For which having an alternate database implementation would not be out of the question.

> 
> At the same time, it would be good to update the documentation for the new packages: https://github.com/ovn-org/libovsdb#documentation
> 

Right, I always forget and I will do that.

> Thank you!

Thanks for the feedback! And reviews are of course very welcome!

jcaamano commented:
I decided to leave the database package split for a later PR so that we can focus on the new model update handling introduced in this PR.

Also added missing documentation.

jcaamano commented:
As for line numbers, approximately 1000 lines of code were removed from `cache/cache.go`, `database/transaction.go`, `ovsdb/updates.go` and `ovsdb/updates2.go`. These were replaced with 680 lines in the updates package. The rest of the changes either more less balance out or are just added tests.

hzhou8 commented:
> > Hi @jcaamano , thanks for the improvements. Could you explain in more detail about the packages created and the motivation?
> > I just looked at the #340, which moved several files of the server package to a new database package, but I didn't understand why was it necessary. It seems those files are still related to server side implementation, so what's the benefit of moving them to a new database package?
> 
>     * For the database package introduced in [server: index uniqueness is a deferred constraint #340](https://github.com/ovn-org/libovsdb/pull/340) , there is a Database interface which the transaction and the server are prepared to use. It does not make sense to have a database implementation in the same package as the server (nor it makes sense to have the database implementation in the same package as the database interface, which I changed in this PR). Someone might want to have a server with a different database implementation.
>

Thank you for the detailed explanation!

If I understand correctly, you are trying to keep the server package only for the RPC interacting with clients, while the new "database" package is to take care of the data management (in memory for now). This leads me to think that the "transaction" module should be part of "database" (which is how the code was organized in PR #340), and should NOT be called directly from the "server" package. But I saw the server package directly calls the Transaction. Should that be decoupled, too? Something like: server -> database -> transaction. Otherwise, it seems not a clear separation between the packages, and they seem to be more suitable as modules in the same package. What do you think?

>     * For the transaction package, we have been considering using it in the client as a validating throw-away transaction, where operations are validated against the client cache before being sent towards the server. Also considering the fact that if someone would have it's own database implementation, it may or may not have it's own transaction implementation as well depending on the desired or needed level of coupling of the transaction with the database.

For the client side transaction validations, I am a little concerned about the cost v.s. the benefit. Integrity violations are considered as rare situations in OVSDB use cases - usually indicating bugs at the client side applications. So I think it is better not doing validations on both client and server. Now that the real OVSDB server does it already, it seems unnecessary at the client side.

>       But we can move this back to the server package for now I guess (let me know if this would be preferred for now) but it was just as easy for me to move it to it's own package.
> 
I guess you meant "the database package". My personal preference is to move it back there (I saw you've already done that when I am replying), because I see it part of the database. Database implementations can also be part of the database package as sub-packages, correct (just like the one "inmemory")? Of course we can always move it out in the future when it is necessary.

>     * The updates package makes sense as model updates are not handled only in the context of a transaction, but also as Update2 modifications are received client side. Previously we had duplicated code in server and client regarding these updates (ApplyModifications & diff).
> 
>     * I don't think that the fact that stuff is only used in the server warrants for everything to be in the same server package anyway. Having separate packages and clear interfaces facilitates reuse and testing so it is nice to push towards an out-of-alpha implementation state for the server as we make changes. Sure, we could actually wait until there is an actual need but it just seemed like an improvement to me in any case so I went for it.
> 
I see that essentially the there are many things in common between the client and the server for data manipulations, so I agree with you that we should reuse code and it is good to have common packages, such as for "updates". I didn't realize this because I didn't pay much attention on the server side in this go project, because I thought it is just for mocked tests.

> 
> All in all, I think they are reasonable decisions. Happy to get specific suggestions on how things could be organized better though.

Now that we are investing more to the server code (although I still have doubts on the long term direction of it, see below), and there are more and more packages relate to the server side, I wonder if we should organize the packages hierarchically. The top level can be client/server/common, and the corresponding packages can be moved under them. However, there can be an unpleasant impact to existing users. Do you have any idea to keep the clear separation between client and server? Or do you think it matter?

> 
> > For the current PR, I got more confused. I haven't looked into deeper yet, but the description claims that it is to reuse common code, yet I am seeing 2000+ code than before with this PR. I appreciate if you could explain your long term plan more clearly to understand such big changes.
> 
> Right, I claimed code reuse, not less code. This added code is contributing to facilitate reuse, currently and for future changes; to fix stuff that was not working correctly on the server side currently; to avoid doing stuff twice at runtime (an updated model is only calculated once, both for the transaction cache and the database cache, instead of twice) probably improving server performance on the way. A single approach converged to being able to to rip all those benefits. But of the 2000+ extra lines, much of them can be attributed to more unit tests as is usually the case. For example, consider that the new updates package has 2500+ lines worth of new tests which are important because it is where we are mainly making changes to data.
> 
I see. Adding new tests of course justifies the added code lines. Big thanks!

> > I appreciate if you could explain your long term plan more clearly to understand such big changes.
> 
> I guess this could have been divided in smaller PRs but the important functional gaps that the server has and the review pace on this project actually favors covering them in less bigger PRs rather than more smaller PRs.
> 
> The short-term plan is to have the server code in a state where we can implement constraint validations reasonably well: we fixed index constraint validation. We want to implement referential integrity validation, maxRows validation, min/max cardinality and other column data type validations, etc. We need these working for testing in downstream projects. The long-term plan is to have the server out of alpha state functionally. The very-long-term plan might be to have the server out of alpha state performance-wise. For which having an alternate database implementation would not be out of the question.
> 
I think I am still confused about the roadmap.

Firstly, even for testing purpose, I am not quite sure about the advantage of using the Go implementation over the real ovsdb-server (C implementation). In this project we still have the integration-test that runs the real ovsdb-server in a container. Is it inconvenient than using the Go implementation?

Secondly, for the client testing purpose, I wonder why we need the advanced validations from the server side. For my understanding it should be easy to write test cases to ensure correctness of the client behavior without requiring the server side validations. I wonder what's your downstream requirement for the testing. Again, if the advanced features are required from the server side, why not using the real ovsdb-server directly?

Finally, for the long-term/very-long-term plan, it seems you have a goal to use the Go based ovsdb-server implementation in production. It would be great to see a better implementation of ovsdb-server, but before moving forward to that goal, I think we should evaluate is there fundamental advantage of rewriting ovsdb-server in Go? Unless it is something hard to be achieved in C or the current OVS community, I don't see how would the Go implementation do better than the C implementation, and I'd rather invest the same effort to improving the C implementation instead. Could you share your view on this?

cc @dave-tucker @amorenoz @trozet @dcbw 

Thanks!

> > At the same time, it would be good to update the documentation for the new packages: https://github.com/ovn-org/libovsdb#documentation
> 
> Right, I always forget and I will do that.
> 
> > Thank you!
> 
> Thanks for the feedback! And reviews are of course very welcome!

jcaamano commented:
> If I understand correctly, you are trying to keep the server package only for the RPC interacting with clients, while the new "database" package is to take care of the data management (in memory for now). This leads me to think that the "transaction" module should be part of "database" (which is how the code was organized in PR https://github.com/ovn-org/libovsdb/pull/340), and should NOT be called directly from the "server" package. But I saw the server package directly calls the Transaction. Should that be decoupled, too? Something like: server -> database -> transaction. Otherwise, it seems not a clear separation between the packages, and they seem to be more suitable as modules in the same package. What do you think?

Agreed. We should expect transactions to have some kind of coupling with the database so it is database implementations which should provide their own transactions. But the transaction implementation as of today might still be useful for other database implementations as long as they comply with the database interface, so while we should allow databases to implement their own transactions, we might not want to preclude them from using the one that is already in place. So certain level of decoupling between the transaction and the database might still be welcomed.

Anyway, what am trying to do in these series of PRs to implement constraint validations in this regard is not flesh out a complete new architecture to have a proper server implementation. I am just trying to express what we already have in place today with what I find a more reasonable package distribution and interfaces so that we can properly re-use code while avoiding circular dependencies.

As mentioned, I removed the database package split I included in this PR initially so we can focus on handling data updates. But I will need it in my next PR to avoid those circular dependencies.

> Now that we are investing more to the server code (although I still have doubts on the long term direction of it, see below), and there are more and more packages relate to the server side, I wonder if we should organize the packages hierarchically. The top level can be client/server/common, and the corresponding packages can be moved under them. However, there can be an unpleasant impact to existing users. Do you have any idea to keep the clear separation between client and server? Or do you think it matter?

I think that perhaps providing some hierarchy for the database package makes sense. A client/server/common hierarchy might not be as useful as I think it's pretty obvious already and we might end up with most of the stuff in common anyway.

> For the client side transaction validations, I am a little concerned about the cost v.s. the benefit. Integrity violations are considered as rare situations in OVSDB use cases - usually indicating bugs at the client side applications. So I think it is better not doing validations on both client and server. Now that the real OVSDB server does it already, it seems unnecessary at the client side.

Certainly the plan would be to make these optional and of course only if the performance cost is reasonable. The main purpose is to verify index uniqueness in the client side. This could be seen as lack of correctness in user implementations but I see out as a way to help out the user. The client does not provide any transactionality or synchronization patterns between user reads & writes, so multithreaded users might find it easier to have this option than to implement their own synchronization mechanisms. Currently we are prepending a wait operation with timeout 0 that fails immediately if the index exists, but this does not work for client side indexes and is less than optimal as well.

> Firstly, even for testing purpose, I am not quite sure about the advantage of using the Go implementation over the real ovsdb-server (C implementation). In this project we still have the integration-test that runs the real ovsdb-server in a container. Is it inconvenient than using the Go implementation?

I would say that for unit tests is it might be better to use the go implementation running in the same test process. In any case, it is already a fact that we are using the this server implementation for ovn-kubernetes unit tests. We don't use it for any other tests.

> Secondly, for the client testing purpose, I wonder why we need the advanced validations from the server side. For my understanding it should be easy to write test cases to ensure correctness of the client behavior without requiring the server side validations. I wonder what's your downstream requirement for the testing. Again, if the advanced features are required from the server side, why not using the real ovsdb-server directly?

This is mainly for referential integrity. The OVSDB RFC describes garbage collection of non-strongly referenced rows from tables that are not part of the root set, as well as garbage collection of weak references to non-existent rows. I don't think that a client relying on this functionality lacks correctness and as such is nice to have in the server implementation and is certainly affecting our testing use of the client.

But another thought of course is that if we are actually using the test server for our unit tests, then it is useful for us as a tool to verify the client correctness and is nice to get a warning sign in the form of a failed operation in the test server.

> Finally, for the long-term/very-long-term plan, it seems you have a goal to use the Go based ovsdb-server implementation in production. It would be great to see a better implementation of ovsdb-server, but before moving forward to that goal, I think we should evaluate is there fundamental advantage of rewriting ovsdb-server in Go? Unless it is something hard to be achieved in C or the current OVS community, I don't see how would the Go implementation do better than the C implementation, and I'd rather invest the same effort to improving the C implementation instead. Could you share your view on this?

I guess I agree with you here, and that is why I used the word `might`. What we are toying with is the idea to use the server as a library (integrating the server in our client process). But before we can even start to play with it we need a reasonable server implementation. In any case, I just laid out ideas and it is far from any formal plan.

What caught me by surprise is the fact that we would need to justify the need to have a correct server implementation. I would have thought that contributions in this regard would be welcomed.

jcaamano commented:
To clarify, I don't think it is an option to use the C implementation for unit tests. Spawning one of those in docker per unit test is not very practical.

Downstream clients might implement their own validation tests/mocks/whatever but wouldn't it be better instead of each downstream user implementing their own version of these, to rather have single test server available that could be not only used in this project but shared by all downstream users to achieve the same thing?

jcaamano commented:
> Certainly the plan would be to make these optional and of course only if the performance cost is reasonable. The main purpose is to verify index uniqueness in the client side. This could be seen as lack of correctness in user implementations but I see out as a way to help out the user. The client does not provide any transactionality or synchronization patterns between user reads & writes, so multithreaded users might find it easier to have this option than to implement their own synchronization mechanisms. Currently we are prepending a wait operation with timeout 0 that fails immediately if the index exists, but this does not work for client side indexes and is less than optimal as well.

To clarify this one. Client users will usually use operation patterns of `createOrUpdate` or `Upsert`. Basically clients will read the cache, and if specific non-indexed or client indexed row is not found, it will insert it, otherwise it will update it. But if the user is multi-threaded, other thread might have inserted the row between the read & write and we will end up with duplicated rows in the DB. Clients might implement their own synchronization mechanisms but these might come at a performance cost compared to been able to detect the problem when it actually happens and retry.

dave-tucker commented:
The motivations for creating a somewhat alpha quality server are here: https://github.com/ovn-org/libovsdb/issues/130
It was a huge pain to try and set up an ovsdb-server, pre-seeded with some data, for each test - this holds true here as well as for downstream projects.

You could do it but you would:
1. Need to spawn ovsdb-server in a docker container (to avoid adding deps on the host, port conflicts etc..)
2. Issue a load of transactions to put the database in the state you want it
3. Do your tests
4. Issue a load of transactions to verify that the server state is the way you want it

Note that 4. could assert against the clients cache - provided that the tables in question are monitored.

With the go-based server implementation the process is much simpler
1. Declarative API to start a server with the desired state (no docker required)
2. Do tests
3. Assert against server state or client state, depending on your monitor implementation.

> Finally, for the long-term/very-long-term plan, it seems you have a goal to use the Go based ovsdb-server implementation in production. It would be great to see a better implementation of ovsdb-server, but before moving forward to that goal, I think we should evaluate is there fundamental advantage of rewriting ovsdb-server in Go?

There are many:
1. Performance - When I last looked, ovsdb-server wasn't multi-threaded and that was hard to add. Concurrency could more easily be added using go (see below).
2. Storage Options - Replacing ovsdb-server storage with purpose built databases was on my TODO list for experimentation - but sadly I've been busy on other projects. It would make ovsdb-server essentially an OVSDB protocol implementation, and remove a lot of complexity (flushing to disk, compaction, clustering etc...). Not only that, but the database can be scaled separately from the protocol layer...
3. JSON performance - A large amount of client/server processing time is spent on JSON processing. There are quite a few more performant options than `encoding/json` which could result in a speedup. As I recall, ovsdb-server has its own JSON Parser and I'm not sure how optimised that is.


I would also add, that since ovsdb-server is a known bottleneck, it's been encouraged to do as much processing client-side as possible. Therefore, having server-like consistency checks that can be validated client-side to avoid flooding the server with unnecessary transactions is a killer feature in my opinion.

hzhou8 commented:
> Agreed. We should expect transactions to have some kind of coupling with the database so it is database implementations which should provide their own transactions. But the transaction implementation as of today might still be useful for other database implementations as long as they comply with the database interface, so while we should allow databases to implement their own transactions, we might not want to preclude them from using the one that is already in place. So certain level of decoupling between the transaction and the database might still be welcomed.

Just to clarify, I am not against decoupling between transaction and the database. They can still be in the same package and sub-packages such as inmemory (and possibly other database implementations) can still share the transaction module.
What I was suggesting was that "server" should NOT directly access "transaction". But I totally understand it is not a responsibility of this PR or the previous one. We can improve it later on.

> 
> I think that perhaps providing some hierarchy for the database package makes sense. A client/server/common hierarchy might not be as useful as I think it's pretty obvious already and we might end up with most of the stuff in common anyway.
> 
I agree if most part will be in "common" then it doesn't make much sense to move. It seems to me at least client, cache, etc. are for the client side, and database, server, etc. are for the server side only. Maybe it is good enough to indicate the client-only and server-only packages in the document, so that people are more clear about the scope of the change they are making.

> To clarify this one. Client users will usually use operation patterns of createOrUpdate or Upsert. Basically clients will read the cache, and if specific non-indexed or client indexed row is not found, it will insert it, otherwise it will update it. But if the user is multi-threaded, other thread might have inserted the row between the read & write and we will end up with duplicated rows in the DB. Clients might implement their own synchronization mechanisms but these might come at a performance cost compared to been able to detect the problem when it actually happens and retry.

Thanks for clarifying and now I see your point. But I don't see how would client side validation solve the multi-threading problem. Libovsdb maintains the cache as read-only from the client, and all updates are from the server notifications, which may be a result of a transaction from another thread or a remote client. The only place to correctly validate the transactions is the server.

> I would say that for unit tests is it might be better to use the go implementation running in the same test process. In any case, it is already a fact that we are using the this server implementation for ovn-kubernetes unit tests. We don't use it for any other tests.
> 
I see, thanks.

> This is mainly for referential integrity. The OVSDB RFC describes garbage collection of non-strongly referenced rows from tables that are not part of the root set, as well as garbage collection of weak references to non-existent rows. I don't think that a client relying on this functionality lacks correctness and as such is nice to have in the server implementation and is certainly affecting our testing use of the client.
> 
This is a great example. Thanks.

> But another thought of course is that if we are actually using the test server for our unit tests, then it is useful for us as a tool to verify the client correctness and is nice to get a warning sign in the form of a failed operation in the test server.
> 
> > Finally, for the long-term/very-long-term plan, it seems you have a goal to use the Go based ovsdb-server implementation in production. It would be great to see a better implementation of ovsdb-server, but before moving forward to that goal, I think we should evaluate is there fundamental advantage of rewriting ovsdb-server in Go? Unless it is something hard to be achieved in C or the current OVS community, I don't see how would the Go implementation do better than the C implementation, and I'd rather invest the same effort to improving the C implementation instead. Could you share your view on this?
> 
> I guess I agree with you here, and that is why I used the word `might`. What we are toying with is the idea to use the server as a library (integrating the server in our client process). But before we can even start to play with it we need a reasonable server implementation. In any case, I just laid out ideas and it is far from any formal plan.
> 
> What caught me by surprise is the fact that we would need to justify the need to have a correct server implementation. I would have thought that contributions in this regard would be welcomed.

Please don't get me wrong. Improvements are of course very welcome. I asked these questions just to better digest such big changes, and to be better prepared for what's coming next. As you have explained, I am more clear at least:
1. Go server implementation has advantages for unit tests
2. To make it work better for the testing purposes, there are still many things to fix/improve - it is definitely not free to get a test server working properly
3. To fix/improve things, we need to refactor the code organization so that we can reuse code

Thanks for your patience :)

hzhou8 commented:
> The motivations for creating a somewhat alpha quality server are here: #130 It was a huge pain to try and set up an ovsdb-server, pre-seeded with some data, for each test - this holds true here as well as for downstream projects.
> 
> You could do it but you would:
> 
>     1. Need to spawn ovsdb-server in a docker container (to avoid adding deps on the host, port conflicts etc..)
> 
>     2. Issue a load of transactions to put the database in the state you want it
> 
>     3. Do your tests
> 
>     4. Issue a load of transactions to verify that the server state is the way you want it
> 
> 
> Note that 4. could assert against the clients cache - provided that the tables in question are monitored.
> 
> With the go-based server implementation the process is much simpler
> 
>     1. Declarative API to start a server with the desired state (no docker required)
> 
>     2. Do tests
> 
>     3. Assert against server state or client state, depending on your monitor implementation.
> 

Thanks @dave-tucker . This makes perfect sense. It justifies the simple Go server implementation, and it is great to see ovn-k8s benefit from this already. At the same time, I think we also need to be careful with big changes. I am not too worried about the changes to the server code, but anything that could impact the client logic needs to be cautious, since it is heavily used in production. It is good to see more unit tests added along with the refactor. More code reviewers may help, too.

> 
> > Finally, for the long-term/very-long-term plan, it seems you have a goal to use the Go based ovsdb-server implementation in production. It would be great to see a better implementation of ovsdb-server, but before moving forward to that goal, I think we should evaluate is there fundamental advantage of rewriting ovsdb-server in Go?
> 
> There are many:
> 
>     1. Performance - When I last looked, ovsdb-server wasn't multi-threaded and that was hard to add. Concurrency could more easily be added using go (see below).
> 
>     2. Storage Options - Replacing ovsdb-server storage with purpose built databases was on my TODO list for experimentation - but sadly I've been busy on other projects. It would make ovsdb-server essentially an OVSDB protocol implementation, and remove a lot of complexity (flushing to disk, compaction, clustering etc...). Not only that, but the database can be scaled separately from the protocol layer...

Great points. I see 1) and 2) are mostly the same thing, and I agree this is a promising approach. Do you have candidates for the backend? ETCD? Anyway, looking forward to seeing this happen.

> 
>     3. JSON performance - A large amount of client/server processing time is spent on JSON processing. There are quite a few more performant options than `encoding/json` which could result in a speedup. As I recall, ovsdb-server has its own JSON Parser and I'm not sure how optimised that is.

This may be true, but like you said, it needs more data to tell.

> 
> I would also add, that since ovsdb-server is a known bottleneck, it's been encouraged to do as much processing client-side as possible. Therefore, having server-like consistency checks that can be validated client-side to avoid flooding the server with unnecessary transactions is a killer feature in my opinion.

I agree with you for "do as much processing client-side as possible" as a principle. But for the client-side validations, I doubt it would really help in performance. From what I've seen, transaction failures happen but are very rare. If it happens frequently then there should be something fundamentally wrong in OVN or in the deployment.

Moreover, no matter how advanced validations are done at the client side, server side has to perform same validations, because client can't validate things between remote clients. 

I agree it would be good to avoid unnecessary transactions even if they are rare, if it doesn't introduce obvious overhead at the client side, and not make the client code unnecessarily complex. (client performance may be not as critical as the server, but for central components such as ovnkube-master, it is also important - ovnkube-master is somehow also one of the bottlenecks)

Again, thank you for the great points!

jcaamano commented:
> Thanks for clarifying and now I see your point. But I don't see how would client side validation solve the multi-threading problem. Libovsdb maintains the cache as read-only from the client, and all updates are from the server notifications, which may be a result of a transaction from another thread or a remote client. The only place to correctly validate the transactions is the server.

In scenarios where the data is written from a single client and transactions are serialized, client side validation might work just as well as the server side validation and may prove more effective than synchronizing multiple big chunks of code between reads & writes. As said, it would be optional if the performance cost when enabled is reasonable to be used in cases where it makes sense.

>  At the same time, I think we also need to be careful with big changes. I am not too worried about the changes to the server code, but anything that could impact the client logic needs to be cautious, since it is heavily used in production.

Well being careful is a given, just as bugs are, but that's not to say we have to stop progress. If you take a closer look at this PR or the previous one, the impact to the client side is pretty limited but I think necessary. The next PRs will be server specific. We have test coverage here and we also try the changes downstream as well as we don't want to bite our own tail.

> Please don't get me wrong. Improvements are of course very welcome. I asked these questions just to better digest such big changes, and to be better prepared for what's coming next.

No worries. Thanks for your patience as well.

flavio-fernandes commented on database/transaction.go:
```diff
@@ -62,45 +63,18 @@ func (t *Transaction) Transact(operations []ovsdb.Operation) ([]*ovsdb.Operation
-					panic(err)
-				}
-			}
+			r, u = t.Insert(&op)
```
👍

flavio-fernandes commented on database/transaction.go:
```diff
@@ -303,292 +244,91 @@ func (t *Transaction) Select(table string, where []ovsdb.Condition, columns []st
-	tableUpdate := make(ovsdb.TableUpdate2)
-
-	rows, err := t.rowsFromTransactionCacheAndDatabase(table, where)
+func (t *Transaction) Update(op *ovsdb.Operation) (ovsdb.OperationResult, *updates.ModelUpdates) {
```
nice!

flavio-fernandes reviewed (commented):
I will need to take a second pass on this change; lots of lines moved.
Assuming what moved is mostly unchanged, I think this is pretty close to a lgtm.
Great cleanup, overall.

jcaamano commented on cache/cache.go:
```diff
@@ -960,53 +960,21 @@ func (t *TableCache) Populate(tableUpdates ovsdb.TableUpdates) error {
-				}
-				t.eventProcessor.AddEvent(deleteEvent, table, oldModel, nil)
-				continue
+		for uuid, row := range tu {
```
I think that specifically the `processing update` log was a bit redundant so that went away. We have the specific insert, update, delete logs in `ApplyCacheUpdate`. Yes, if we have an unexpected error, those would not be logged but the error instead would be logged. Just as any other error we would have had before we reached Populate.

Buuut I can add it back here actually.

jcaamano commented on cache/cache.go:
```diff
@@ -960,53 +960,21 @@ func (t *TableCache) Populate(tableUpdates ovsdb.TableUpdates) error {
-				continue
+		for uuid, row := range tu {
+			update := updates.ModelUpdates{}
+			current := tCache.cache[uuid]
```
That clone is now happening in `update.AddRowUpdate`. And that's the idea, the clones happen inside the updates package, as that package is the one that is actually modifying the models.

jcaamano commented on cache/cache_test.go:
```diff
@@ -2900,3 +2649,130 @@ func BenchmarkRowsByCondition(b *testing.B) {
+		expected model.Model
+	}{
+		{
+			name: "create",
```
Sure! Will add one. We should not get it, though.

jcaamano commented:
@flavio-fernandes PTAL

Added the missing log and the test you suggested.

hzhou8 commented:
Thanks @flavio-fernandes . Sorry for my slow response. I will try to complete my review this week.

hzhou8 commented:
/lgtm
Thanks @jcaamano

jcaamano commented:
Addressing concerns on performance impacts in the client code path, these changes only affect Populate/Populate2. We have a couple of benchmarks on those methods:

With this PR:
```
goos: linux
goarch: amd64
pkg: github.com/ovn-org/libovsdb/cache
cpu: AMD EPYC-Rome Processor
BenchmarkPopulate2UpdateArray-4              106          10871796 ns/op         3902213 B/op      82653 allocs/op
BenchmarkPopulate2SingleModify-4              43          28169521 ns/op        13291977 B/op     228197 allocs/op
PASS
ok      github.com/ovn-org/libovsdb/cache       4.938s
```

With the code base:
```
goos: linux
goarch: amd64
pkg: github.com/ovn-org/libovsdb/cache
cpu: AMD EPYC-Rome Processor
BenchmarkPopulate2UpdateArray-4               60          18998553 ns/op         6238954 B/op     234004 allocs/op
BenchmarkPopulate2SingleModify-4              39          28613911 ns/op        13306334 B/op     228197 allocs/op
PASS
ok      github.com/ovn-org/libovsdb/cache       3.755s
```

Note that `BenchmarkPopulate2SingleModify` is not in the code base but is equivalent to the `BenchmarkTableCacheApplyModificationsSet` we had in the code base. I just backported `BenchmarkPopulate2SingleModify` to the cxode base to be able to compare.

cc @dcbw
